File: release_artifacts/examples/cloud/firebase_example.ts
/**
 * @file firebase_example.ts
 * @brief TypeScript example usage of Firebase integration with Satox SDK
 * @copyright Copyright (c) 2025 Satoxcoin Core Developers
 * @license MIT License
 */

import { DatabaseManager, DatabaseType } from '../../bindings/typescript/src/database';
import * as fs from 'fs';

async function main() {
    console.log('=== Satox SDK Firebase Integration Example (TypeScript) ===');
    
    try {
        // Initialize database manager
        const dbManager = new DatabaseManager();
        
        // Load Firebase configuration
        const configData = fs.readFileSync('config/cloud/firebase_config.json', 'utf8');
        const config = JSON.parse(configData);
        
        // Connect to Firebase
        const connectionId = await dbManager.connect(DatabaseType.FIREBASE, config);
        if (!connectionId) {
            console.error('Error: Failed to connect to Firebase');
            return;
        }
        
        console.log(`Successfully connected to Firebase with ID: ${connectionId}`);
        
        // Example: Firestore operations
        console.log('\n--- Firestore Operations ---');
        
        // Create a document
        const createDocQuery = JSON.stringify({
            operation: 'set',
            collection: 'users',
            document_id: 'user123',
            data: {
                name: 'John Doe',
                email: 'john@example.com',
                age: 30,
                created_at: '2025-01-20T10:00:00Z'
            }
        });
        
        const createResult = await dbManager.executeQuery(connectionId, createDocQuery);
        if (createResult.success) {
            console.log('Document created successfully');
        } else {
            console.error('Error creating document:', createResult.error);
        }
        
        // Get a document
        const getDocQuery = JSON.stringify({
            operation: 'get',
            collection: 'users',
            document_id: 'user123'
        });
        
        const getResult = await dbManager.executeQuery(connectionId, getDocQuery);
        if (getResult.success) {
            console.log('Document retrieved successfully');
            console.log('Document data:', JSON.stringify(getResult.rows, null, 2));
        } else {
            console.error('Error retrieving document:', createResult.error);
        }
        
        // Update a document
        const updateDocQuery = JSON.stringify({
            operation: 'update',
            collection: 'users',
            document_id: 'user123',
            data: {
                age: 31,
                updated_at: '2025-01-20T11:00:00Z'
            }
        });
        
        const updateResult = await dbManager.executeQuery(connectionId, updateDocQuery);
        if (updateResult.success) {
            console.log('Document updated successfully');
        } else {
            console.error('Error updating document:', updateResult.error);
        }
        
        // Query documents
        const queryDocQuery = JSON.stringify({
            operation: 'query',
            collection: 'users',
            data: {
                where: {
                    field: 'age',
                    operator: '>=',
                    value: 25
                }
            }
        });
        
        const queryResult = await dbManager.executeQuery(connectionId, queryDocQuery);
        if (queryResult.success) {
            console.log('Query executed successfully');
            console.log('Query results:', JSON.stringify(queryResult.rows, null, 2));
        } else {
            console.error('Error querying documents:', queryResult.error);
        }
        
        // Delete a document
        const deleteDocQuery = JSON.stringify({
            operation: 'delete',
            collection: 'users',
            document_id: 'user123'
        });
        
        const deleteResult = await dbManager.executeQuery(connectionId, deleteDocQuery);
        if (deleteResult.success) {
            console.log('Document deleted successfully');
        } else {
            console.error('Error deleting document:', deleteResult.error);
        }
        
        // Example: Transaction
        console.log('\n--- Transaction Example ---');
        const transactionQueries = [
            JSON.stringify({
                operation: 'set',
                collection: 'users',
                document_id: 'user1',
                data: { name: 'Alice', email: 'alice@example.com' }
            }),
            JSON.stringify({
                operation: 'set',
                collection: 'users',
                document_id: 'user2',
                data: { name: 'Bob', email: 'bob@example.com' }
            })
        ];
        
        const transactionResult = await dbManager.executeTransaction(connectionId, transactionQueries);
        if (transactionResult.success) {
            console.log('Transaction executed successfully');
        } else {
            console.error('Error executing transaction:', transactionResult.error);
        }
        
        // Disconnect
        const disconnectResult = await dbManager.disconnect(connectionId);
        if (disconnectResult) {
            console.log('Disconnected from Firebase');
        } else {
            console.error('Error disconnecting');
        }
        
    } catch (error) {
        console.error('Exception:', error);
    }
    
    console.log('=== Example completed ===');
}

main().catch(console.error);

File: release_artifacts/examples/cloud/firebase_example.cpp
/**
 * @file firebase_example.cpp
 * @brief C++ example usage of Firebase integration with Satox SDK
 * @copyright Copyright (c) 2025 Satoxcoin Core Developers
 * @license MIT License
 */

#include "satox/core/database_manager.hpp"
#include "satox/core/cloud/firebase_config.hpp"
#include <nlohmann/json.hpp>
#include <iostream>
#include <fstream>
#include <chrono>
#include <thread>

using json = nlohmann::json;
using satox::core::DatabaseManager;
using satox::core::DatabaseType;
using satox::core::cloud::FirebaseConfig;

int main() {
    std::cout << "=== Satox SDK Firebase Integration Example (C++) ===" << std::endl;
    
    try {
        // Initialize database manager
        DatabaseManager db_manager;
        
        // Load Firebase configuration
        std::ifstream config_file("config/cloud/firebase_config.json");
        if (!config_file.is_open()) {
            std::cerr << "Error: Could not open firebase_config.json" << std::endl;
            return 1;
        }
        
        json config_json;
        config_file >> config_json;
        config_file.close();
        
        // Connect to Firebase
        std::string connection_id = db_manager.connect(DatabaseType::FIREBASE, config_json);
        if (connection_id.empty()) {
            std::cerr << "Error: Failed to connect to Firebase" << std::endl;
            std::cerr << "Last error: " << db_manager.getLastError() << std::endl;
            return 1;
        }
        
        std::cout << "Successfully connected to Firebase with ID: " << connection_id << std::endl;
        
        // Example: Create a collection
        json create_result;
        std::string create_collection_query = R"(
            CREATE COLLECTION IF NOT EXISTS users
        )";
        
        if (db_manager.executeQuery(connection_id, create_collection_query, create_result)) {
            std::cout << "Collection created successfully" << std::endl;
        } else {
            std::cout << "Error creating collection: " << create_result["error"] << std::endl;
        }
        
        // Example: Insert data
        json insert_result;
        std::string insert_query = R"(
            INSERT INTO users (username, email, created_at) 
            VALUES ('john_doe', 'john@example.com', NOW())
        )";
        
        if (db_manager.executeQuery(connection_id, insert_query, insert_result)) {
            std::cout << "Data inserted successfully" << std::endl;
        } else {
            std::cout << "Error inserting data: " << insert_result["error"] << std::endl;
        }
        
        // Example: Query data
        json query_result;
        std::string select_query = "SELECT * FROM users WHERE username = 'john_doe'";
        
        if (db_manager.executeQuery(connection_id, select_query, query_result)) {
            std::cout << "Query executed successfully" << std::endl;
            if (query_result.contains("rows") && query_result["rows"].is_array()) {
                std::cout << "Found " << query_result["rows"].size() << " users:" << std::endl;
                for (const auto& row : query_result["rows"]) {
                    std::cout << "  - Username: " << row["username"]
                              << ", Email: " << row["email"] << std::endl;
                }
            }
        } else {
            std::cout << "Error querying data: " << query_result["error"] << std::endl;
        }
        
        // Example: Update data
        json update_result;
        std::string update_query = R"(
            UPDATE users SET email = 'john.updated@example.com' WHERE username = 'john_doe'
        )";
        
        if (db_manager.executeQuery(connection_id, update_query, update_result)) {
            std::cout << "Data updated successfully" << std::endl;
        } else {
            std::cout << "Error updating data: " << update_result["error"] << std::endl;
        }
        
        // Example: Real-time features
        std::cout << "\n=== Real-time Features ===" << std::endl;
        std::cout << "Note: Real-time features require WebSocket support" << std::endl;
        std::cout << "This would typically involve:" << std::endl;
        std::cout << "1. Setting up WebSocket connection" << std::endl;
        std::cout << "2. Subscribing to collection changes" << std::endl;
        std::cout << "3. Handling real-time events" << std::endl;
        
        // Example: Authentication
        std::cout << "\n=== Authentication Features ===" << std::endl;
        std::cout << "Firebase provides built-in authentication:" << std::endl;
        std::cout << "1. Email/Password authentication" << std::endl;
        std::cout << "2. Google Sign-In" << std::endl;
        std::cout << "3. Facebook Sign-In" << std::endl;
        std::cout << "4. Anonymous authentication" << std::endl;
        
        // Disconnect
        if (db_manager.disconnect(connection_id)) {
            std::cout << "Successfully disconnected from Firebase" << std::endl;
        } else {
            std::cout << "Error disconnecting: " << db_manager.getLastError() << std::endl;
        }
        
    } catch (const std::exception& e) {
        std::cerr << "Exception: " << e.what() << std::endl;
        return 1;
    }
    
    std::cout << "\n=== Example completed successfully ===" << std::endl;
    return 0;
}
